torch/csrc/autograd/engine.cpp

```diff
@@ -69,6 +73,47 @@ struct ReadyQueue {
   FunctionTask pop();
 };
 
+// Note [Reentrant backwards]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~
+// To understand the reentrant backwards problem, we have to notice two
+// aspects of how the autograd engine is implemented today:
+//
+//  1. When you call Engine::execute(), you want to block until autograd
+//  execution finishes so that you can get the final result variables
+//  of the backwards pass.
+//
+//  2. Autograd execution operates by having a *single* worker assigned
+//  for any given task (based on the device the task it will run on),
+//  which is responsible for executing the contents of the task.
+//
+// The problem is, suppose that you call backward() inside of a worker
+// thread.  By property (1), we're supposed to block until the nested task
+// finishes.  However, by property (2), this worker thread is on the
+// hook for processing the tasks assigned to it; we better not block,
+// because then all of our backward executions (including the one we
+// just started) will deadlock!
+//
+// Here's our cunning idea: instead of blocking, just get back to work
+// on whatever task queue you should have been working on previously
+// (this is saved via the TLS variable worker_device)!  There are
+// simply two things you have to arrange for:
+//
+//  - We have to promptly kick ourselves out of the thread_main() loop
+//    when our graph_task complete, because we need to unblock the
+//    parent function tasks that started the reentrant execution in
+//    the first place.  This is why thread_main() takes an optional
+//    graph_task as input.
+//
+//  - When we finish a graph, we have to make sure we wake up the worker
+//    thread so that it actually has a chance to exit the thread_main()
+//    loop.  Thus the faffing about in thread_main() after
+//    evaluate_function() completes.
+
+
+// There is one GraphTask per execution of backward(); it holds the
+// metadata for an overall backwards execution.  There may be multiple
```

Comment:
First two sentences say pretty much the same thing. Let's limit them to just "GraphTask holds metadata needed for a single execution of backward()"